server/src/main/java/org/opensearch/indices/tiering/HotToWarmTieringService.java

```diff
@@ -0,0 +1,875 @@
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.Priority;
+import org.opensearch.common.annotation.ExperimentalApi;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.lifecycle.AbstractLifecycleComponent;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.common.util.FeatureFlags;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.index.Index;
+import org.opensearch.core.index.shard.ShardId;
+import org.opensearch.index.IndexModule;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.opensearch.cluster.metadata.IndexMetadata.INDEX_NUMBER_OF_REPLICAS_SETTING;
+import static org.opensearch.cluster.metadata.IndexMetadata.TIERING_CUSTOM_KEY;
+import static org.opensearch.index.IndexModule.INDEX_STORE_LOCALITY_SETTING;
+import static org.opensearch.index.IndexModule.INDEX_TIERING_STATE;
+
+/**
+ * Service responsible for tiering indices from hot to warm
+ * @opensearch.experimental
+ */
+public class HotToWarmTieringService extends AbstractLifecycleComponent implements ClusterStateListener {
+    private static final Logger logger = LogManager.getLogger(HotToWarmTieringService.class);
+    protected final ClusterService clusterService;
+    protected final IndexNameExpressionResolver indexNameExpressionResolver;
+    protected final AllocationService allocationService;
+    // map to store all the in progress tiering requests
+    private final Map<String, TieringRequestContext> requestUuidToRequestContext = new ConcurrentHashMap<>();
+    // map to store index info of all the indices under-going hot to warm tiering
+    private final Map<Index, IndexTieringInfo> indexTieringInfoMap = new ConcurrentHashMap<>();
+    public static final String HOT_TO_WARM_START_TIME = "hot_to_warm_start_time";
+    public static final String HOT_TO_WARM_END_TIME = "hot_to_warm_end_time";
+
+    @Inject
+    public HotToWarmTieringService(
+        Settings settings,
+        ClusterService clusterService,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        AllocationService allocationService
+    ) {
+        super();
+        this.clusterService = clusterService;
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+        this.allocationService = allocationService;
+
+        if (DiscoveryNode.isClusterManagerNode(settings) && FeatureFlags.isEnabled(FeatureFlags.TIERED_REMOTE_INDEX)) {
+            clusterService.addListener(this);
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // TODO add handling for master switch, dangling indices, master reload
+        if (event.routingTableChanged()) {
```

Comment:
Is this race condition scenario possible?
```
Lets say before allocation service picked the indices for shard reroute or allocation, some other index got deleted.  
so this cluster change event will be called where updateIndexShardStatus method will update the tiering state to FAILED however tiering/relocation haven't started yet.
```

Reply:
updated the logic